OAPE-25: Customizing control plane machine names

chiragkyal opened the pull request:
This enhancement proposal

- Provides a more flexible way to name Control Plane Machines managed by the `ControlPlaneMachineSet`. 
- Introduces the `MachineNamePrefix` field allowing custom prefixes to be used.
- This feature is gated behind the `CPMSMachineNamePrefix` feature gate.

Part of [OAPE-25](https://issues.redhat.com/browse/OAPE-25)

chiragkyal commented on enhancements/machine-api/control-plane-machine-set.md:
```diff
@@ -113,6 +113,17 @@ on the spec and status structures defined below.
+	// and an index of the machine. The prefix must not end with a special character,
+	// and only lowercase alphanumeric characters and hyphens ('-') are allowed, with hyphens permitted only
+	// between characters. The prefix must be between 1 and 32 characters in length.
+	// +kubebuilder:validation:Pattern=`^[a-z0-9]([-a-z0-9]*[a-z0-9])?$`
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=32
```
This was inspired by the [ClusterID](https://github.com/openshift/api/blob/a2817b89f7e0989016967055a142d5e88bed18de/config/v1/types_cluster_version.go#L51-L56) validation which follows the RFC4122 UUID value. However I'm open to suggestions if it needs to be modified.

JoelSpeed replied:
The name prefix should be validated as a dns1123 subdomain name, normally the max length is 253 chars, but, we need to suffix this string with 5 random chars and the index (e.g. `-abcde-0`) so we should reduce the max length to meet that need, 245 chars should be the max length.

You can use CEL to validate the subdomain name format now

```
// +kubebuilder:validation:XValidation:rule="format.dns1123label.validate(self).hasValue()",messageExpression="format.dns1123label.validate(self).value().join(\"\\n\")"
```

chiragkyal replied:
I think the prefix validation should work for all cloud providers. I was checking GCP's naming convention https://cloud.google.com/compute/docs/naming-resources#resource-name-format, and it looks like dns1123 subdomain check will fail there.

JoelSpeed replied:
The name here is the name of the Machine object, which can be DNS1123 subdomain, it is up to the provider implementation to make sure that the name of the Machine can be translated to the cloud provider naming schemes. The GCP provider for example will have logic to convert the name should it not be compliant.

I'd prefer to start with a kube compatible naming scheme, and change while in TP if we find issues

Can you please check if the installer does any validation of the cluster ID, that way we can see what is there already in the fleet

chiragkyal replied:
>The name here is the name of the Machine object, which can be DNS1123 subdomain, it is up to the provider implementation to make sure that the name of the Machine can be translated to the cloud provider naming schemes. The GCP provider for example will have logic to convert the name should it not be compliant.
>
> I'd prefer to start with a kube compatible naming scheme, and change while in TP if we find issues

That sounds good to me. We can start with DNS1123 subdomain validation while in TP. 

So, should we use `format.dns1123label` or `format.dns1123Subdomain` for doing the CEL validation?

> Can you please check if the installer does any validation of the cluster ID, that way we can see what is there already in the fleet

I checked that installer code, and started a thread with the team : https://redhat-internal.slack.com/archives/C68TNFWA2/p1732184902373169

Summary: Installer uses the ClusterName which is then transformed to create `ClusterID` of length 27 and follow `DNS1123Subdomain`

JoelSpeed replied:
Use subdomain for the formatting

chiragkyal replied:
I tried using the CEL rule in the API. But it wasn't able to recognize `format`. Do we need to import some package?

```
compilation failed: ERROR: <input>:1:14: undeclared reference to 'format' (in container '')
```

chiragkyal replied:
Another question: Is it acceptable to keep the index as a single-digit (e.g. `-abcde-0`)? I think in that case (having `245` characters as max length prefix) a cluster can have a maximum of `9` control plane nodes, right?

JoelSpeed replied:
OpenShift 4 only supports 3 control plane nodes, it will not support more than that in the near future, and even if it did, I would not expect us to ever support more than 5 control plane nodes (note that's the most we've ever supported across any OpenShift)

chiragkyal replied:
Okay, thanks for the confirmation. Then we are good with that validation.

chiragkyal commented:
/assign @JoelSpeed

JoelSpeed reviewed (commented):
What happens if I have an existing cluster and the Machines are not named based on my prefix, because I have no prefix, and then I add the prefix? Is the desire that this would replace the entire control plane just to rename some machines, or would it fix the names on the next spec change?

chiragkyal commented:
> What happens if I have an existing cluster and the Machines are not named based on my prefix, because I have no prefix, and then I add the prefix? Is the desire that this would replace the entire control plane just to rename some machines, or would it fix the names on the next spec change?

With my WIP implementation, I tested this scenario. I edited the `ControlPlaneMachineSetSpec` and added the prefix. The operator did not rename all the machines, just after the spec change. However, once I deleted one of the existing machines, a new machine was created with the new name convention (with prefix). I hope this behaviour is what we're expecting.

huali9 commented on enhancements/machine-api/control-plane-machine-set.md:
```diff
@@ -754,6 +770,19 @@ discover during testing something that is currently unknown to us. If we discove
+
+* It must be between 1 and 32 characters in length.
+* It must only contain lowercase alphanumeric characters and hyphens (`-`).
+* Hyphens are only permitted between characters (i.e. the prefix cannot start or end with a hyphen).
```
In the previous part https://github.com/openshift/enhancements/pull/1714/files#diff-167c14da636a95920dc7f720fd0d0be7490305b02914e4113ee9aebf2f51a3bdR119-R122 , it says consisting of `lowercase alphanumeric characters, '-', or '.'`  `between 1 and 245 characters in length.` but here is `1 and 32 characters` `lowercase alphanumeric characters and hyphens (`-`)`, which is inconsistent, is this right?

huali9 replied:
Got it ,thank you!
LGTM

JoelSpeed commented on enhancements/machine-api/control-plane-machine-set.md:
```diff
@@ -983,7 +1019,10 @@ TBD
 
-TBD
+##### MachineNamePrefix Functionality
+- Removal (enabled by default) of `CPMSMachineNamePrefix` feature gate usage.
```
Nit, `Removal` is not the right word here, prefer to use `Promote` the feature gate

chiragkyal replied:
Thank you, updated

JoelSpeed reviewed (commented):
One nit, otherwise LGTM

@damdo did you have anything you needed to raise?

JoelSpeed commented:
/approve
/assign @damdo

chiragkyal commented:
@damdo, can we merge the EP if there are no concerns?

JoelSpeed commented:
/lgtm
